src/topology/metric_space/pi_lp.lean

```diff
@@ -0,0 +1,369 @@
+  $$
+  when `p ≥ 1`. We prove it in `sum_rpow_minkowski_of_nonneg` when
+  `a i` and `b i` are nonnegative reals, and we also give versions over `nnreal` and `ennreal`.
+
+## Implementation notes
+
+The Hölder and Minkowski inequalities have more general versions (for the integral over a general
+measure space). This file only contains the finset version, which is the one needed to define finite
+products of metric spaces. The finset version could be derived from the general version (which is
+not formalized yet).
+-/
+
+open real set
+open_locale big_operators uniformity topological_space
+
+noncomputable theory
+
+variables {ι : Type*}
+
+namespace real
+
+/-- Two real exponents `p, q` are conjugate if they are `> 1` and satisfy the equality
+`1/p + 1/q = 1`. This condition shows up in many theorems in analysis, notably related to `L^p`
+norms. -/
+structure is_conjugate_exponent (p q : ℝ) : Prop :=
+(one_lt : 1 < p)
+(inv_add_inv_conj : 1/p + 1/q = 1)
+
+/-- The conjugate exponent of `p` is `q = p/(p-1)`, so that `1/p + 1/q = 1`. -/
+def conjugate_exponent (p : ℝ) : ℝ := p/(p-1)
+
+/- Register several non-vanishing results following from the fact that `p` has a conjugate exponent
+`q`: many computations using these exponents require clearing out denominators, which can be done
+with `field_simp` given a proof that these denominators are non-zero, so we record the most usual
+ones. -/
+lemma is_conjugate_exponent.pos {p q : ℝ} (h : p.is_conjugate_exponent q) :
+  0 < p :=
+lt_trans zero_lt_one h.one_lt
+
+lemma is_conjugate_exponent.ne_zero {p q : ℝ} (h : p.is_conjugate_exponent q) :
+  p ≠ 0 :=
+ne_of_gt h.pos
+
+lemma is_conjugate_exponent.sub_one_ne_zero {p q : ℝ}
+  (h : p.is_conjugate_exponent q) : p - 1 ≠ 0 :=
+sub_ne_zero_of_ne (ne_of_gt h.one_lt)
+
+lemma is_conjugate_exponent.one_div_pos {p q : ℝ} (h : p.is_conjugate_exponent q) :
+  0 < 1/p :=
+one_div_pos_of_pos h.pos
+
+lemma is_conjugate_exponent.one_div_ne_zero {p q : ℝ} (h : p.is_conjugate_exponent q) :
+  1/p ≠ 0 :=
+ne_of_gt (h.one_div_pos)
+
+lemma is_conjugate_exponent_iff {p q : ℝ} (h : 1 < p) :
+  p.is_conjugate_exponent q ↔ q = p/(p-1) :=
+begin
+  have A : 1/p + 1/q = 1 ↔ q = 1/(1-1/p) := calc
+    1/p + 1/q = 1 ↔ 1/q = 1-1/p : eq_sub_iff_add_eq'.symm
+    ... ↔ q = 1/(1-1/p) : by { simp only [one_div_eq_inv], rw inv_eq_iff, exact eq_comm },
+  split,
+  { assume H,
+    rw A.1 H.inv_add_inv_conj,
+    field_simp [H.ne_zero] },
+  { assume H,
+    refine ⟨h, _⟩,
+    rw H,
+    field_simp [ne_of_gt (lt_trans zero_lt_one h)] }
+end
+
+lemma is_conjugate_exponent_conjugate_exponent {p : ℝ} (h : 1 < p) :
+  p.is_conjugate_exponent (conjugate_exponent p) :=
+(is_conjugate_exponent_iff h).2 rfl
+
+namespace is_conjugate_exponent
+
+variables {p q : ℝ} (h : p.is_conjugate_exponent q)
+include h
+
+lemma conj_eq : q = p/(p-1) :=
+(is_conjugate_exponent_iff h.one_lt).1 h
+
+protected lemma symm : q.is_conjugate_exponent p :=
```

Comment:
```suggestion
@[symm] protected lemma symm : q.is_conjugate_exponent p :=
```
Is there any reason to use `namespace` here but not earlier?

Reply:
BTW, could you please move `real.is_conjugate_exponent` to a separate file&PR? Then we'll merge it, I'll reformulate `real.young_inequality` using this definition, and use it to golf the proof of Hölder inequality.

Reply:
Done in #2992. I'll rebase this one when #2992 is merged, and you can golf as much as you want.